hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderSelectBuilder.java

```diff
@@ -224,32 +251,47 @@ private LoaderSelectBuilder(
 		this.cachedDomainResult = cachedDomainResult;
 		this.numberOfKeysToLoad = numberOfKeysToLoad;
 		this.loadQueryInfluencers = loadQueryInfluencers;
+
+		this.forceIdentifierSelection = determineWhetherToForceIdSelection( numberOfKeysToLoad, restrictedParts );
+		this.entityGraphTraversalState = determineGraphTraversalState( loadQueryInfluencers );
+
+		this.lockOptions = lockOptions != null ? lockOptions : LockOptions.NONE;
+		this.jdbcParameterConsumer = jdbcParameterConsumer;
+	}
+
+	private static boolean determineWhetherToForceIdSelection(int numberOfKeysToLoad, List<ModelPart> restrictedParts) {
 		if ( numberOfKeysToLoad > 1 ) {
-			forceIdentifierSelection = true;
+			return true;
 		}
-		else {
-			for ( ModelPart restrictedPart : restrictedParts ) {
-				if ( restrictedPart instanceof ForeignKeyDescriptor ) {
-					forceIdentifierSelection = true;
-				}
+
+		if ( restrictedParts.size() == 1 ) {
+			final ModelPart restrictedPart = restrictedParts.get( 0 );
+			if ( Objects.equals( restrictedPart.getPartName(), NaturalIdMapping.PART_NAME ) ) {
```

Comment:
Given `NaturalIdMapping.PART_NAME` is not null, it seems we can replace the above with
```
if ( NaturalIdMapping.PART_NAME.equals( restrictedPart.getPartName() ) {
```
to save some unnecessary null checking in `Objects.equals()`.